### PR TITLE
Adds evilbit plugin and inital documentation

### PR DESCRIPTION
**Pull request by ana-cc**

### CODE REVIEW

File: pathspider/plugins/evilbit.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+from scapy.all import IP  # pylint: disable=E061
+from scapy.all import IPv6  # pylint: disable=E061
+from scapy.all import UDP  # pylint: disable=E061
+from scapy.all import TCP  # pylint: disable=E061
+from scapy.all import DNS  # pylint: disable=E061
+from scapy.all import DNSQR  # pylint: disable=E061
+from scapy.all import RandShort  # pylint: disable=E061
```
Comment (irl) [inline pathspider/plugins/evilbit.py]: All the disable=E601 need to be changed to "disable=no-name-in-module" as the pylint on Travis doesn't like the old-style disables.

File: pathspider/plugins/evilbit.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+from pathspider.chains.tcp import TCPChain
+from pathspider.chains.evil import EvilChain
+from pathspider.chains.base import Chain
+from pathspider.chains.tcp import TCP_SYN, TCP_SA
+
+
+class EvilBit(ForgeSpider, PluggableSpider):
+
+    name = "evilbit"
+    description = "Evil bit connectivity testing (RFC 3514)"
```
Comment (irl) [inline pathspider/plugins/evilbit.py]: We don't put references in the description, we put them in the documentation.

File: pathspider/plugins/evilbit.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    #version = pathspider.__version__
```
Comment (irl) [inline pathspider/plugins/evilbit.py]: Uncomment this, I don't remember where the version is now but setup.py will tell you.

File: pathspider/plugins/evilbit.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                  DNS(qd=DNSQR(qname=job['domain'])))
+        if ':' in job['dip']:
+            ip = IPv6(src=self.source[1], dst=job['dip'])
+        else:
+            ip = IP(src=self.source[0], dst=job['dip'])
+        if seq == 1:
+            ip.flags = 'evil'
+        return ip/l4
+
+    def combine_flows(self, flows):
```
Comment (irl) [inline pathspider/plugins/evilbit.py]: Can you refactor this to have less branches? Make use of a private function (starting with a underscore) if needed.

File: pathspider/plugins/evilbit.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            if not flows[0]['dns_response_valid'] and flows[1][
+                    'dns_response_valid']:
+                conditions.append('evilbit.connectivity.transient')
+            else:
+                conditions.append('evilbit.connectivity.offline')
+
+        return conditions
+
+    @classmethod
+    def register_args(cls, subparsers):
```
Comment (irl) [inline pathspider/plugins/evilbit.py]: This whole function is redundant, as you'll inherit this from ForgeSpider anyway. The timeout doesn't appear to be used either, which ForgeSpider doesn't have by default.

File: pathspider/chains/evil.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+from pathspider.chains.tcp import TCP_SYN
+from pathspider.chains.base import Chain
+
+class EvilChain(Chain):
+
+    """
+    This flow analysis chain records details of the evil bit in the IP header
+
+     ``evil_bit_syn_fwd``  boolean   True if the evil bit was set in the IP header for a TCP SYN packet in the
```
Comment (irl) [inline pathspider/chains/evil.py]: This should be a reStructuredText table.

File: pathspider/chains/evil.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        :type rec: dict
+        :param ip: the IPv4 packet that was observed to be part of this flow
+        :type ip: plt.ip
+        :param rev: True if the packet was in the reverse direction, False if
+                    in the forward direction
+        :type rev: bool
+        :return: Always ``True``
+        :rtype: bool
+        """
+        return self._eb_extract(rec, ip, rev)
```
Comment (irl) [inline pathspider/chains/evil.py]: Evil Bit in itself doesn't indicate any state for the connection, these should just be allowed to be merged after the timeout. This allows the chain to be used by Sync/Desync plugins as well as by Forge plugins.

File: pathspider/chains/evil.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+
+    def ip6(self, rec, ip, rev):
```
Comment (irl) [inline pathspider/chains/evil.py]: Just remove this whole function, it's redundant. You can add the note that IPv6 is not supported to the class docstring.

File: pathspider/chains/evil.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        """
+
+        return True	
+    
+    def _eb_extract(self, rec, ip, rev):
+        evil = ip.has_rf
+            
+        if ip.tcp:
+            if ip.tcp.flags & TCP_SYN == TCP_SYN:
+                rec['evil_bit_syn_rev' if rev else 'evil_bit_syn_fwd'] = evil
```
Comment (irl) [inline pathspider/chains/evil.py]: Use "evilbit_" not "evil_bit_" to match the other naming in other chains.

File: pathspider/plugins/evilbit.py (hunk 1 above)
Comment (ana-cc) [inline pathspider/plugins/evilbit.py]: Done, no issues.

File: pathspider/plugins/evilbit.py (hunk 2 above)
Comment (ana-cc) [inline pathspider/plugins/evilbit.py]: Done, no issues.

File: pathspider/plugins/evilbit.py (hunk 3 above)
Comment (ana-cc) [inline pathspider/plugins/evilbit.py]: Found it, no issues.

File: pathspider/plugins/evilbit.py (hunk 5 above)
Comment (ana-cc) [inline pathspider/plugins/evilbit.py]: Additional arguments I want to pass(like connect_supported = ["tcpsyn", "dnsudp"]) do not appear to register on the command line if I remove this function.

File: pathspider/chains/evil.py (hunk 6 above)
Comment (ana-cc) [inline pathspider/chains/evil.py]: Done, no issues.

File: pathspider/chains/evil.py (hunk 7 above)
Comment (ana-cc) [inline pathspider/chains/evil.py]: Sorry, I'm not sure I understand what needs changing.

File: pathspider/chains/evil.py (hunk 8 above)
Comment (ana-cc) [inline pathspider/chains/evil.py]: Done, no issues.

File: pathspider/chains/evil.py (hunk 9 above)
Comment (ana-cc) [inline pathspider/chains/evil.py]: Done.

File: pathspider/plugins/evilbit.py (hunk 5 above)
Comment (irl) [inline pathspider/plugins/evilbit.py]: In which case, that's a bug in forgespider. Can you file an issue for that and point out in that issue that the register_args function needs to be removed from evil bit when it's fixed.

File: pathspider/chains/evil.py (hunk 7 above)
Comment (irl) [inline pathspider/chains/evil.py]: You have a few options. One would be to move the contents of _eb_extract into the ip4 function, as this is the only place it is used. This is probably what I would do.

The point is that you return the return value of _eb_extract when it is is better to explicitly return True here, the return status shouldn't depend on _eb_extract which isn't immediately clear when you are returning the return value of that function.

**Comment by irl** [comment]
You will need to update the test_plugin_load tests to include evilbit. You will also need to include some test cases for the evil bit chain.

File: pathspider/plugins/evilbit.py (hunk 4 above)
Comment (ana-cc) [inline pathspider/plugins/evilbit.py]: Done.

File: pathspider/plugins/evilbit.py (hunk 5 above)
Comment (ana-cc) [inline pathspider/plugins/evilbit.py]: Support for different type of connection arguments missing from forgespider. New plugins that use forgespider can't take advantage of the inherited 'connect_supported' argument that desync spider has.
I can just fix this

File: pathspider/chains/evil.py (hunk 7 above)
Comment (ana-cc) [inline pathspider/chains/evil.py]: I had a look and saw the silly code. Have done what you suggested, thanks!

**Pull request closed without merging**